README.md
![themify](https://i.imgur.com/JZyjWm6.png)

[![Build Status](https://travis-ci.org/datorama/themify.svg?branch=master)](https://travis-ci.org/datorama/themify)
[![All Contributors](https://img.shields.io/badge/all_contributors-2-orange.svg?style=flat-square)](#contributors)
[![commitizen](https://img.shields.io/badge/commitizen-friendly-brightgreen.svg?style=flat-square)]()
[![PRs](https://img.shields.io/badge/PRs-welcome-brightgreen.svg?style=flat-square)]()
[![styled with prettier](https://img.shields.io/badge/styled_with-prettier-ff69b4.svg?style=flat-square)](https://github.com/prettier/prettier)
[![semantic-release](https://img.shields.io/badge/%20%20%F0%9F%93%A6%F0%9F%9A%80-semantic--release-e10079.svg?style=flat-square)](https://github.com/semantic-release/semantic-release)

> *CSS Themes Made Easy*

Themify lets you manage your application’s themes in realtime, using a robust solution that’s easily configurable.

Themify is a PostCSS plugin that generates your theme during the build phase.
The main concept behind it is to provide two palettes, one light and one dark (resembles the inverse of the light palette).

Under the hood, `themify` will replace your CSS colors with CSS variables, and also take care to provide a fallback for unsupported browsers (such as IE11).

[Introducing Themify: CSS Themes Made Easy](https://engineering.datorama.com/introducing-themify-css-themes-made-easy-669b7ecb8720)

## 🤓 Features

* 🖌 **Light &  Dark palettes** - define your theme using a simple JSON format

* 🎨 **Replace your colors in runtime** - provide your clients with  **white-labeling** capabilities. Let them choose their own colors and replace them instantly

* :pencil2: **Use it inside your CSS** - use your theme directly in your SASS files, no JavaScript is required

* 🏃 **Runtime replacement** - change the active palette at runtime, either for the entire application or for a specific HTML container

* 🔥 **Legacy Browser Support** - support for all major browsers including IE11

## Installation
`npm install @datorama/themify --save`

## Usage

#### Options

|Input|Type|Default|Description|
|---|---|---|---|
|createVars|boolean|`true`|Determines whether CSS variables are automatically generated. This should kept as true, unless you want to inject them yourself.|
|palette|`{light: [key: string]: string, dark: [key: string]: string}`|`{}`|Palette colors|
|classPrefix|string|`''`|A class prefix to append to each generated theme class.|
|screwIE11|boolean|`true`|Whether to generate a fallback for legacy browsers that do not supports CSS Variables.|
|fallback|`{cssPath: string \| null, dynamicPath: string \| null}`|`{}`|`cssPath`: An absolute path to the fallback CSS. <br>`dynamicPath`: An absolute path to the fallback JSON.|

#### Add themify to your build pipe: 

```js
const themifyOptions = {  
  palette : {
    light: {
      'primary-100': '#f2f2f4',
      'primary-200': '#cccece',
      'accent-100': '#e6f9fc',
      'accent-200': '#96e1ed'
    },
    dark: {
      'primary-100': '#505050',
      'primary-200': '#666a6b',
      'accent-100': '#096796',
      'accent-200': '#0a87c6'
    }
  },
  screwIE11 : false,  
  fallback : {  
    cssPath : './dist/theme_fallback.css', // use checksum 
    dynamicPath: './dist/theme_fallback.json'  
  }  
};
```

##### Gulp

```js
gulp.src('./main.scss')  
    .pipe(postcss([  
	initThemify(themifyOptions),  
	sass(),  
	themify(themifyOptions)  
     ]))  
    .pipe(rename("bundle.css"))  
    .pipe(gulp.dest('dist'));
```

##### Webpack

```js
const isProd = process.env.ENV === 'production';
const basePath = isProd ? './dist' : './src';
const cssPath = `${basePath}/theme_fallback.css`;
const dynamicPath = `${basePath}/theme_fallback.json`;

{
  test: /\.scss$/,
  use: [{loader: "style-loader"}].concat(getLoaders())
}
      
const getLoaders = () => [{
  loader: "css-loader"
},
  {
    loader: 'postcss-loader',
    options: {
      ident: 'postcss2',
      plugins: () => [
        require('@datorama/themify').themify(themifyOptions)
      ]
    }
  },
  {
    loader: "sass-loader"
  },
  {
    loader: 'postcss-loader',
    options: {
      ident: 'postcss1',
      plugins: () => [
        require('@datorama/themify').initThemify(themifyOptions)
      ]
    }
  }
]
```


#### Add themify to SASS

In order to use the `themify` function and other SASS helpers, you need to import the `themify` library from your main SASS file:

```sass
@import 'node_modules/datorama/themify/dist/themify';
```

The themify function receives as parameters the name of the color defined in the palette map and an optional opacity parameter. Themify will generate CSS selectors for each palette — one for the light and one for the dark.

```scss
.my-awesome-selector {
  // color-key: a mandatory key from your palette. For example: primary-100
  // opacity: an optional opacity. Valid values between 0 - 1. Defaults 1.
  background-color: themify(color-key, opacity);

  // Define a different color for dark and light.
  color: themify((dark: color-key-1, light: color-key-2));
}
```

#### Basic usage

```scss
button {
  background-color: themify(primary-100);
  color: themify(accent-200);
  &:hover {
    background-color: themify(primary-100, 0.5);	
  }
}
```

The above example will produce the following CSS:

```css
.dark button, button {
  background-color: rgba(var(--primary-100), 1);
  color: rgba(var(--accent-200), 1);
}
.dark button:hover, button:hover {
  background-color: rgba(var(--primary-100), 0.5);	
}
```

And the following fallback for IE11:

```css
button {
  background-color: #f2f2f4;
  color: #666a6b;
}
 
.dark button {
  background-color: #505050;
  color: #0a87c6;
}

button:hover {
  background-color: rgba(242, 242, 244, 0.5);	
}

.dark button:hover {
  background-color: rgba(80, 80, 80, 0.5);	
}
```

#### A different color for each palette

Sometimes we need more control over the colors so it's possible to specify explicitly one color for **light** and another color for **dark**:

```scss
button {
  background-color: themify((dark: primary-100, light: accent-200));
}
```

The above example will produce the following CSS:

```css
.button {
  background-color: rgba(var(--accent-200), 1);
}

.dark button {
  background-color: rgba(var(--primary-100), 1);
}
```

#### Advanced usage

`themify` can be combined with every valid CSS:

```scss
button {
  border: 1px solid themify(primary-100);
  background: linear-gradient(themify(accent-200), themify(accent-100));
}
```

Even in your animations:

```scss
.element {  
  animation: pulse 5s infinite;  
}  
  
@keyframes pulse {
  0% {  
    background-color: themify(accent-100);  
  }  
  100% {  
    background-color: themify(accent-200);  
  }  
}
```

#### Runtime replacement

First, we'll create our own theme service.

```ts
import {loadCSSVariablesFallback, replaceColors, Theme} from '@datorama/themify/utils';
const palette = require('path_to_my_json_pallete');

/** fallback for CSS variables support */
const themeCSSFallback = 'path/theme_fallback.css';
const themeJSONFallback = 'path/theme_fallback.json';

export class MyThemeService {
	
  constructor(){
    /**
    * load the CSS fallback file, in case the browser do not support CSS Variables.
    * Required only if you set screwIE11 option to false. 
    *
    * callback - load event for the CSS file
    */
    loadCSSVariablesFallback(themeCSSFallback, callback);	
  }

  /**  
   * Replace the theme colors at runtime 
   * @param partialTheme a partial theme configuration.  
   */
   setColors(partialTheme: Theme){
     replaceColors(themeJSONFallback, partialTheme, palette);
   }

}
```

Now let's use this service in our web application:

```ts
const themeService = new MyThemeService();

/** replace the colors at runtime **/
themeService.setColors({
  light: {
    'primary-100': '#0c93e4'
  }
});

```

 
#### Changing the active palette

In order to switch between the dark and light palettes, simply add the appropriate class to the desired HTML element.

```scss
p {
  /** #96e1ed in light and #0a87c6 in dark */
  color: themify(accent-200);
}
```

```html
<p>I'm from the light palette</p>
<div class="dark">
  <p>I'm from the dark palette</p>	
</div>
```
### Theme class helpers
You can take advantage of your themes not just in your CSS, but also directly in your HTML, by generating a CSS class for each color you define.

In order to achieve this, use the `generateThemeHelpers` mixin, and pass the CSS properties you want to generate. For example:

```scss
// generates the following predefined classes, for each color  
$themeRules: (  
  'color',  
  'border-top-color',  
  'border-bottom-color',  
  'border-right-color',  
  'border-left-color',  
  'background-color',  
  'fill',  
  'stroke',  
  // PSEUDO_CLASSES  
  'color:h:f:a:vi'  
);  
@include generateThemeHelpers($themeRules);
```

This will generate the following CSS:

```css
.dark .primary-100-color, .primary-100-color {  
  color: rgba(var(--primary-100), 1)  
}

.dark .primary-200-color, .primary-200-color {  
  color: rgba(var(--primary-100), 1)  
}

.dark .primary-100-color\:vi:visited, .primary-100-color\:vi:visited {  
  color: rgba(var(--primary-100), 1)  
}
```
and so on..

As you see, you can pass any CSS property, including pseudo classes.
The following SASS map details the pseudo class keys and their values:

```sass
$PSEUDO_CLASSES: (  
  ':a': ':active',  
  ':c': ':checked',  
  ':d': ':default',  
  ':di': ':disabled',  
  ':e': ':empty',  
  ':en': ':enabled',  
  ':fi': ':first',  
  ':fc': ':first-child',  
  ':fot': ':first-of-type',  
  ':fs': ':fullscreen',  
  ':f': ':focus',  
  ':h': ':hover',  
  ':ind': ':indeterminate',  
  ':ir': ':in-range',  
  ':inv': ':invalid',  
  ':lc': ':last-child',  
  ':lot': ':last-of-type',  
  ':l': ':left',  
  ':li': ':link',  
  ':oc': ':only-child',  
  ':oot': ':only-of-type',  
  ':o': ':optional',  
  ':oor': ':out-of-range',  
  ':ro': ':read-only',  
  ':rw' : ':read-write',  
  ':req': ':required',  
  ':r': ':right',  
  ':rt' : ':root',  
  ':s': ':scope',  
  ':t' : ':target',  
  ':va': ':valid',  
  ':vi': ':visited'  
);
```

<br>
Now you can use the generated CSS classes directly in your HTML:

```html
<a class="primary-100-color primary-200-color:a">
	The default color is primary-100
	The active color will be primary-200
</a>
```

## Known issues
- We discovered that Safari doesn't support the following syntax when it comes to borders with CSS variables:
```css
/** This will NOT work */
border: 1px solid themify(primary-100);

/** This will work */
border-color: themify(primary-100);

/** This will work */
border: themify(primary-100) 1px solid;
```

- Safari doesn't support box-shadow.

## Contributors

Thanks goes to these wonderful people ([emoji key](https://github.com/kentcdodds/all-contributors#emoji-key)):

<!-- ALL-CONTRIBUTORS-LIST:START - Do not remove or modify this section -->
<!-- prettier-ignore -->
| [<img src="https://avatars1.githubusercontent.com/u/6745730?v=4" width="100px;"/><br /><sub><b>Netanel Basal</b></sub>](https://www.netbasal.com)<br />[📖](https://github.com/datorama/themify/commits?author=NetanelBasal "Documentation") [💻](https://github.com/datorama/themify/commits?author=NetanelBasal "Code") [🤔](#ideas-NetanelBasal "Ideas, Planning, & Feedback") | [<img src="https://avatars2.githubusercontent.com/u/78281?v=4" width="100px;"/><br /><sub><b>bh86</b></sub>](https://github.com/bh86)<br />[📖](https://github.com/datorama/themify/commits?author=bh86 "Documentation") [💻](https://github.com/datorama/themify/commits?author=bh86 "Code") [🤔](#ideas-bh86 "Ideas, Planning, & Feedback") |
| :---: | :---: |
<!-- ALL-CONTRIBUTORS-LIST:END -->

This project follows the [all-contributors](https://github.com/kentcdodds/all-contributors) specification. Contributions of any kind welcome!

## License

Apache &copy; [datorama](https://github.com/datorama)


CHANGELOG.md
<a name="1.0.5"></a>
## [1.0.5](https://github.com/datorama/themify/compare/v1.0.4...v1.0.5) (2018-05-24)


### Bug Fixes

* **package:** remove hextorgba package ([9da3fce](https://github.com/datorama/themify/commit/9da3fce))

<a name="1.0.4"></a>
## [1.0.4](https://github.com/datorama/themify/compare/v1.0.3...v1.0.4) (2018-05-23)


### Bug Fixes

* **npm:** add readme file ([cb3aa35](https://github.com/datorama/themify/commit/cb3aa35))

<a name="1.0.3"></a>
## [1.0.3](https://github.com/datorama/themify/compare/v1.0.2...v1.0.3) (2018-05-23)


### Bug Fixes

* **npm:** add missing details ([491ed9f](https://github.com/datorama/themify/commit/491ed9f))

<a name="1.0.2"></a>
## [1.0.2](https://github.com/datorama/themify/compare/v1.0.1...v1.0.2) (2018-05-13)


### Bug Fixes

* **tests:** upgrade node-sass to latest version ([b4e66f5](https://github.com/datorama/themify/commit/b4e66f5))

<a name="1.0.1"></a>
## [1.0.1](https://github.com/datorama/themify/compare/v1.0.0...v1.0.1) (2018-05-10)


### Bug Fixes

* **playground:** fix native variables assertion ([1505b58](https://github.com/datorama/themify/commit/1505b58))
* **themify:** css fallback - minification and at-rule fix ([c0b06a8](https://github.com/datorama/themify/commit/c0b06a8))

<a name="1.0.0"></a>
# 1.0.0 (2018-05-07)


### Features

* **themify:** init ([728c429](https://github.com/datorama/themify/commit/728c429))
* **themify:** init ([eeb947d](https://github.com/datorama/themify/commit/eeb947d))


playground/helpers.js
let JSONFallbackCache;

/**
 *
 * @param path
 */
function loadJSON(url, cb) {
  const req = new XMLHttpRequest();
  req.overrideMimeType('application/json');
  req.open('GET', url, true);
  req.onload = function() {
    cb(JSON.parse(req.responseText));
  };
  req.send(null);
}

/**
 *
 * @param path
 */
function loadCSS(path) {
  const head = document.getElementsByTagName('head')[0];
  const style = document.createElement('link');
  style.href = path;
  style.id = 'themify-fallback';
  style.type = 'text/css';
  style.rel = 'stylesheet';
  head.appendChild(style);
}

/**
 *
 * @param style
 */
function injectStyle(style) {
  var node = document.createElement('style');
  node.id = 'themify';
  console.log('injecting style', style);
  node.innerHTML = style;
  document.head.appendChild(node);
}

/**
 *
 * .dark {
 *   --primary-100: 30, 24, 33;
 * }
 *
 * :root {
 *   --primary-100: 22, 21, 22;
 * }
 *
 * @param customTheme
 * @returns {string}
 */
function generateNewVariables(customTheme) {
  // First, we need the variations [dark, light]
  const variations = Object.keys(customTheme);
  return variations.reduce((finalOutput, variation) => {
    // Next, we need the variation keys [primary-100, accent-100]
    const variationKeys = Object.keys(customTheme[variation]);

    const variationOutput = variationKeys.reduce((acc, variable) => {
      const value = normalizeColor(customTheme[variation][variable]);
      return (acc += `--${variable}: ${value};`);
    }, '');

    return (finalOutput += `${variation === 'light' ? ':root' : '.' + variation}{${variationOutput}}`);
  }, '');
}

/**
 *
 * @returns {boolean}
 */
function hasNativeCSSProperties() {
  return false;
  const opacity = '1';
  const el = document.head;
  let hasNativeCSSProperties;

  // Setup CSS properties.
  el.style.setProperty('--test-opacity', opacity);
  el.style.setProperty('opacity', 'var(--test-opacity)');

  // Feature detect then remove all set properties.
  hasNativeCSSProperties = window.getComputedStyle(el).opacity == opacity;
  el.style.setProperty('--test-opacity', '');
  el.style.opacity = '';

  return hasNativeCSSProperties;
}

/**
 * Load the CSS fallback file on load
 */
function loadCSSVariablesFallback(fallbackPath) {
  if (!hasNativeCSSProperties()) {
    loadCSS(fallbackPath);
  }
}

/**
 *
 * @param customTheme
 */
function replaceColors(fallbackJSONPath, customTheme) {
  if (customTheme) {
    if (hasNativeCSSProperties()) {
      const newColors = generateNewVariables(customTheme);
      injectStyle(newColors);
    } else {
      const replace = JSONFallback => {
        JSONFallbackCache = JSONFallback;
        handleUnSupportedBrowsers(customTheme, JSONFallbackCache);
      };
      if (JSONFallbackCache) {
        replace(JSONFallbackCache);
      } else {
        loadJSON(fallbackJSONPath, replace);
      }
    }
  }
}

/**
 *
 * @param customTheme
 */
function handleUnSupportedBrowsers(customTheme, JSONFallback) {
  const themifyRegExp = /%\[(.*?)\]%/gi;
  const merged = mergeDeep(palette, customTheme);

  let finalOutput = Object.keys(customTheme).reduce((acc, variation) => {
    let value = JSONFallback[variation].replace(themifyRegExp, (occurrence, value) => {
      const [variation, variable, opacity] = value.replace(/\s/g, '').split(',');
      const color = merged[variation][variable];
      const normalized = hexToRGB(color, opacity);
      return normalized;
    });

    return (acc += value);
  }, '');

  injectStyle(finalOutput);
}

/**
 * Omit the rgb and braces from rgb
 * rgb(235, 246, 244) => 235, 246, 244
 * @param rgb
 * @returns {string}
 */
function normalizeRgb(rgb) {
  return rgb.replace('rgb(', '').replace(')', '');
}

/**
 *
 * @param color
 * @returns {*}
 */
function normalizeColor(color) {
  if (isHex(color)) {
    return normalizeRgb(hexToRGB(color));
  }

  if (isRgb(color)) {
    return normalizeRgb(color);
  }

  return color;
}

/**
 *
 * @param color
 * @returns {boolean}
 */
function isHex(color) {
  return color.indexOf('#') > -1;
}

/**
 *
 * @param color
 * @returns {boolean}
 */
function isRgb(color) {
  return color.indexOf('rgb') > -1;
}

/**
 *
 * @param hex
 * @param alpha
 * @returns {string}
 */
function hexToRGB(hex, alpha = false) {
  const r = parseInt(hex.slice(1, 3), 16);
  const g = parseInt(hex.slice(3, 5), 16);
  const b = parseInt(hex.slice(5, 7), 16);
  if (alpha) {
    return `rgba(${r}, ${g}, ${b}, ${alpha})`;
  }
  return `rgb(${r}, ${g}, ${b})`;
}

var palette = {
  light: {
    'primary-700': '#303030',
    'primary-600': '#383838',
    'primary-500': '#505050',
    'primary-400': '#666a6b',
    'primary-300': '#9ca0a0',
    'primary-200': '#cccece',
    'primary-100': '#f2f2f4',
    'primary-50': '#f8f8f9',
    'primary-0': '#ffffff',
    'accent-700': '#096796',
    'accent-600': '#0a87c6',
    'accent-500': '#04a2d6',
    'accent-400': '#00bee8',
    'accent-300': '#4cd1ef',
    'accent-200': '#96e1ed',
    'accent-100': '#e6f9fc'
  },
  dark: {
    'primary-700': '#ffffff',
    'primary-600': '#f8f8f9',
    'primary-500': '#f2f2f4',
    'primary-400': '#cccece',
    'primary-300': '#9ca0a0',
    'primary-200': '#666a6b',
    'primary-100': '#505050',
    'primary-50': '#383838',
    'primary-0': '#303030',
    'accent-700': '#e6f9fc',
    'accent-600': '#96e1ed',
    'accent-500': '#4cd1ef',
    'accent-400': '#00bee8',
    'accent-300': '#04a2d6',
    'accent-200': '#0a87c6',
    'accent-100': '#096796'
  }
};

/**
 *
 * @param target
 * @param sources
 * @returns {*}
 */
function mergeDeep(target, ...sources) {
  if (!sources.length) return target;
  const source = sources.shift();

  if (isObject(target) && isObject(source)) {
    for (const key in source) {
      if (isObject(source[key])) {
        if (!target[key]) Object.assign(target, { [key]: {} });
        mergeDeep(target[key], source[key]);
      } else {
        Object.assign(target, { [key]: source[key] });
      }
    }
  }

  return mergeDeep(target, ...sources);
}

/**
 *
 * @param item
 * @returns {*|boolean}
 */
function isObject(value) {
  return Object.prototype.toString.call(value) === '[object Object]';
}
